### PR TITLE
MCD: Possible test pattern

### PR DESCRIPTION
**Pull request by ashcrow**
Looking through the MCD code I found trying to write unittests a bit harder than normal. One pattern I've seen in the past is to pass functions (or structs, etc..) that have network requests/side effects as part of the design of the system. This unbinds said side effects at the unit level and, in theory, can provide a foundation for different execution paths chosen at runtime. 

**Note**: `New` probably wasn't example to use to test with since it's pretty simple, but I think the idea still is shown.

This is a quick example of:

- making a type of `NodeAnnotationLoader`
- renaming `loadNodeAnnotations` to `LoadNodeAnnotations`
- passing `LoadNodeAnnotations` to `New`
- optional passing of `login1.Conn` to `New`. If passed `nil` one will be created inside of `New`.
- creating a mocked function for testing that also implements `NodeAnnotationLoader` for unittesting

```console
$ cd pkg/daemon/
$ go test -v -run ^TestNew$
=== RUN   TestNew
--- PASS: TestNew (0.00s)
PASS
ok      github.com/openshift/machine-config-operator/pkg/daemon 0.009s
```

### CODE REVIEW

**Comment by ashcrow** [comment]
Huh. Tests pass locally, fail in prow. Looking at this now.

**Comment by ashcrow** [comment]
Ah this makes sense. Because the unittest is happening within a container and is not mounting in `/var/run/dbus/system_bus_socket` the test fails. Being that I'm on a desktop it was able to see the file. Adding another function input.

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ import (
 	drain "github.com/wking/kubernetes-drain"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
+	kubernetescorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
```
[the remark this answers is not in the document]
Comment (ashcrow) [reply pkg/daemon/daemon.go]: Will update

**Comment by sdemos** [comment]
I'm a little on the fence with this approach. It gets the job done, and it does nicely move side-effect-related code to particular functions that will naturally be more controlled with this model.

The problem for me is that it seems unwieldy to have all of those types of functions have their own defined function prototype and be function pointer arguments into every function that needs them. 

I think I'm partial to moving all related types of side-effect code that needs to be mocked to various interfaces, that then can have a real client and a mock client. then we just have a different client for each different kind of part (eg what we vaguely have now, but it's not well adhered to).

thoughts?

**Comment by ashcrow** [comment]
> The problem for me is that it seems unwieldy to have all of those types of functions have their own defined function prototype and be function pointer arguments into every function that needs them.

Absolutely agree :-/

> I think I'm partial to moving all related types of side-effect code that needs to be mocked to various interfaces, that then can have a real client and a mock client. then we just have a different client for each different kind of part (eg what we vaguely have now, but it's not well adhered to).
>
> thoughts?

I think that's a good approach as well. The downside is that it's not just clients, but also functions that have side effects themselves. If we can broaden it to be more than just what is considered a client I think it would be doable. Or if we define client as anything interfacing with an external entity (IE: file system client, kubernetes client, login1 client, etc..).

**Comment by sdemos** [comment]
> Or if we define client as anything interfacing with an external entity (IE: file system client, kubernetes client, login1 client, etc..).

this is more concretely what I meant. I vaguely tried writing the original agent with this approach in mind, but didn't stick to it very well, and it's only gone further away from this approach afaict. it seems like the downside is that it would be a more significant refactor.

**Comment by ashcrow** [comment]
@sdemos post #14 merger I'll take a look at making clients so we can later implement mocks for tests.

**Comment by ashcrow** [comment]
@sdemos during a meeting I threw [this](https://gist.github.com/ashcrow/bd1f3df68262d6a07c5db57304c2d6e3) together. In your opinion do you think each test suite should set up it's own mock to hang off clients, or should mocks be "smart" pass/fail based on some kind of input ... either to the client or via variable setting?

**Comment by ashcrow** [comment]
See https://github.com/openshift/machine-config-operator/issues/101

**Pull request closed without merging**